### ISSUE TITLE
Disowned process loses access to keyring

### ISSUE THREAD

**Issue by maxammann** [google/fscrypt#146]
Hello,
I experimented with fscrypt and process which have been disowned in the shell over SSH.

A user on my system has an encrypted home which gets unlocked through pam. There is also an other encrypted directory named "/var/enc":

```
Session Keyring
 355931798 --alswrv   1000  1000  keyring: _ses
1003096489 --alswrv   1000 65534   \_ keyring: _uid.1000
  29857105 --alsw-v   1000  1000       \_ logon: ext4:a463da817a0cc6a8 # home
 628287813 --alsw-v      0     0       \_ logon: ext4:c7cdddc43fadc5e6 # /var/enc
```
## Steps to reproduce
I have do not have the settings `drop_caches` and `lock_policies` set in the pam module.

I have two SSH sessions open:
* One session as root user
* One session as user max (uid: 1000)

I do now the following in the SSH session of max:
* `python2 /var/enc/pyload-stable/pyLoadCore.py`
* Press CTRL+Z
* `bg`
* `disown`

As soon as I close the SSH session of max I get file access errors. See the following strace log:
```
mkdir("scripts/pyload_start", 0777)     = -1 ENOKEY (Required key not available)
stat64("scripts/pyload_restart", 0xbecdf990) = -1 ENOENT (No such file or directory)
stat64("scripts", {st_mode=S_IFDIR|0775, st_size=4096, ...}) = 0
```

```
Traceback (most recent call last):
  File "/var/pyload/pyload-stable/module/lib/bottle.py", line 733, in _handle
  File "/var/pyload/pyload-stable/module/lib/bottle.py", line 1448, in wrapper
  File "/var/pyload/pyload-stable/module/web/api_app.py", line 48, in call_api
  File "/var/pyload/pyload-stable/module/lib/beaker/session.py", line 554, in __getattr__
  File "/var/pyload/pyload-stable/module/lib/beaker/session.py", line 550, in _session
  File "/var/pyload/pyload-stable/module/lib/beaker/session.py", line 119, in __init__
  File "/var/pyload/pyload-stable/module/lib/beaker/session.py", line 205, in invalidate
  File "/var/pyload/pyload-stable/module/lib/beaker/session.py", line 211, in load
  File "/var/pyload/pyload-stable/module/lib/beaker/container.py", line 537, in __init__
    OpenResourceNamespaceManager.__init__(self, namespace)
  File "/var/pyload/pyload-stable/module/lib/beaker/container.py", line 116, in __init__
    self.access_lock = self.get_access_lock()
OSError: [Errno 126] Required key not available: './tmp/container_file_lock
```

If the errors do not occur right away then I can do the following in the root SSH session:
* sync
* echo "2" > /proc/sys/vm/drop_caches

This is not needed though.

## Expected behavior

The python application should still have access to the keyring/keys.

## My guess

I was not yet able to reproduce this with an other python application.
Therefore I think that the python application makes weird stuff with processing.
**Is there a way to verify which keyring a process has access to?**

**Comment by ebiggers** [comment]
> I have do not have the settings drop_caches and lock_policies set in the pam module.

Do you mean you have them set, or you don't have them set?

Also, which Linux distribution is this on?

You can execute the `keyctl show` command to check which keyrings a process is subscribed to.  Try executing it from the application.

**Comment by josephlr** [comment]
@maxammann I think the issue here is that when your user logs out of the last session, all of the directories for your user get locked by the PAM module. If your PAM `session` options don't have `lock_policies` set, then this is an issue.

As @ebiggers mentioned, this might be an issue in that a disowned process might not have access to your user keying. Let me know what `keyctl show` gives when run from the disowned process.

**Comment by josephlr** [comment]
I just confirmed that doing these steps and disowning the process keeps access to the appropriate session and user keyrings. So the only things I can think of are:
  
  - `lock_policies` is set in your PAM module session configuration.
  - There's a bug in the `lock_policies` flag.

**Comment by maxammann** [comment]
> Also, which Linux distribution is this on?

The distribution is Armbian which is based on Ubuntu Bionic. The kernel version is the following:
```
Linux max-nas 4.19.63-mvebu #5.91 SMP Wed Jul 31 15:58:43 CEST 2019 armv7l armv7l armv7l GNU/Linux
```

> Do you mean you have them set, or you don't have them set?

The two flags are disabled. The pam config looks like this:
```
session optional        pam_fscrypt.so debug
```

@josephlr I also was not able to create a python programm which has the same problem. I will debug this further by executing `keyctl` from the python2 process.


Thanks for your support! With that information I can debug further.

**Comment by maxammann** [comment]
As soon as the ZSH process stops and the ssh session ends the output of `keyctl show` is empty:

```
Session Keyring
 714911291 --alswrv   1000  1000  keyring: _ses
 327407076 --alswrv   1000 65534   \_ keyring: _uid.1000
 231764531 --alsw-v   1000  1000       \_ logon: ext4:a463da817a0cc6a8
 502940595 --alsw-v   1000  1000       \_ logon: ext4:c7cdddc43fadc5e6
Session Keyring
 714911291 --alswrv   1000  1000  keyring: _ses
 327407076 --alswrv   1000 65534   \_ keyring: _uid.1000
 231764531 --alsw-v   1000  1000       \_ logon: ext4:a463da817a0cc6a8
 502940595 --alsw-v   1000  1000       \_ logon: ext4:c7cdddc43fadc5e6
Session Keyring
Session Keyring
Session Keyring
```

I suppose the process is somehow forking or creating a new subprocess which does not get the same access to the keyring. I think the next step would be to create a minimal python programm which makes this reproducable.

**Comment by maxammann** [comment]
The process tree looks like this:

![2019-09-16-20:30:36-screenshot](https://user-images.githubusercontent.com/905221/64983493-e04d6600-d8c0-11e9-8bef-5e9cf3547da6.png)

(strace does not change the behavior)

**Comment by josephlr** [comment]
@maxammann thanks for this! It really helps narrow down the problem. Here's the gist of what the problem seems to be:

- Kernel keyrings have garbage collection, so that when no processes are using them, they are closed/destroyed
- fscrypt puts it's keys in the user keyring (`_uid.1000` in your example), which is then linked by the session keyring (`_ses` in your example). This session keyring is the one that is then searched when you try to access an encrypted file.
- It looks like when a process gets disowned, the kernel stops counting that process as "using" the session keyring (for some reason). But it's initially OK, as the session keyring is still being used by ZSH/SSH.
- However, when ZSH/SSH exits, the reference count on `_ses` drops to 0, and it's removed, and thus access to the user keyring and fscrypt keys are also removed.

I'm actually not sure if there is a way to fix this, we might just have to update the documentation to say that the disowned processes don't play well with kernel keyrings. I see two potential ways to work around this issue:

1. Explicitly link the user keyring into your process keyring.
2. Once support for Linux 5.3. lands, use the new fscrypt kernel API (which avoids any use of kernel keyrings at all). Unfortunately, this will require updating your kernel (and updating fscrypt).

 @ebiggers any ideas here?

@maxammann, to confirm this, can you run the same output as before, but also print out the contents of /`proc/key-users` and `/proc/keys` in your loop?

**Comment by maxammann** [comment]
@josephlr I think you are right, that's what I suspected. I also think that the problem is buried in the kernel.

1. I successfully created a new session and linked that with my user keyring. I am not sure though how to link a user and process keyring.

2. So with 5.3 there will be a kernel API which handles the access to files for fscrypt? Can you link me to something where I can track it and potentially contribute?

I will send the output of `/proc/key-users` and `/proc/keys` soon.

**Comment by ebiggers** [comment]
Forked processes keep the same session keyring unless they override it, so it's unclear why your process suddenly loses access to it.  Is there any chance your process is calling some library which creates a new session keyring?  Other options are that something else on the system could be invalidating, revoking, or expiring the session keyring, or changing the permissions on it, or that the process is changing its UID causing itself to lose access.  Does the output you showed include both stdout and stderr, or just stdout?  

Anyway, I'd much rather that people migrate to the new filesystem-level encryption keyring, rather than debugging all these crazy issues with the process-subscribed keyrings.  The kernel patches will likely be merged for 5.4 this week ([link to pull request](https://lkml.kernel.org/lkml/20190916044258.GA8269@sol.localdomain/)), and I'll soon be opening a pull request here for the [userspace patches](https://github.com/ebiggers/fscrypt/commits/fscrypt-key-mgmt-improvements).  If you want, you can help test the patches.  For now you'd need to build the kernel from [this branch](https://git.kernel.org/pub/scm/fs/fscrypt/fscrypt.git/log/) and `fscrypt` from [this branch](https://github.com/ebiggers/fscrypt/tree/fscrypt-key-mgmt-improvements).

**Comment by maxammann** [comment]
Alright I debugged this a little bit further as there could have been a bug in the application I was running.

I wrote a small script to write the process information available:
```python
def f():
    while True:
        import subprocess
        import time
        import pwd
        import datetime
        f = open("/tmp/keyctl-test.txt", "a")

        def w(c):
            f.write(str(c) + os.linesep)

        w(datetime.datetime.now())
        ..
        w(os.getpid())
        w(os.getppid())
        ...
        w('')
        subprocess.call(["/bin/keyctl", "show"], stdout=f)
        time.sleep(1)

if __name__ == '__main__':
    t = threading.Thread(target=f)
    t.start()
    t.join()
```

As soon as I close the SSH session the ppid changes to 1 which is the `/sbin/init` process. This process is owned by `root`.
The output of keyctl is correct as long as the SSH session is running.

Now here is the important part: As soon as the SSH session is closed the `_sess` keyring is 'destroyed' and the process looses access to the decryption key!

If I join the `_uid.1000` other keyring in the shell session and therefore leave the `_sess` keyring then the process is allowed to access the keyring:
```
keyctl session _uid.1000
```

After joining that session the ppid changes to 1 after leaving the SSH session and the python process still has access to the decryption key!

So there is no bug there. The reason for this behavior is the destruction of the `_sess` keyring! This can be confirmed by executing `keyctl show` in different SSH sessions.

@ebiggers I hope I will find time to test the patches in the 5.4 kernel and to setup a test VM with the self-compiled kernel.

Cheers! Closing this as it is not a bug ;)
